File: VulkanGameEngine/GameObjectSystem.cpp
#include "GameObjectSystem.h"
#include "LevelSystem.h"

GameObjectSystem gameObjectSystem = GameObjectSystem();

GameObjectSystem::GameObjectSystem()
{
}

GameObjectSystem::~GameObjectSystem()
{
}

void GameObjectSystem::CreateGameObject(const String& name, const Vector<ComponentTypeEnum>& gameObjectComponentTypeList, VkGuid vramId, vec2 objectPosition)
{
	GameObjectID id;
	id.id = GameObjectMap.size() + 1;
	GameObjectMap[id] = GameObject(id);

	for (auto component : gameObjectComponentTypeList)
	{
		switch (component)
		{
			case kTransform2DComponent: Transform2DComponentMap[id] = Transform2DComponent(objectPosition); break;
			case kInputComponent: InputComponentMap[id] = InputComponent(id); break;
			case kSpriteComponent: levelSystem.SpriteList[id] = Sprite(id, vramId); break;
		}
	}
}

void GameObjectSystem::CreateGameObject(const String& gameObjectPath, const vec2& gameObjectPosition)
{
	GameObjectID id;
	id.id = GameObjectMap.size() + 1;
	GameObjectMap[id] = GameObject(id);

	nlohmann::json json = Json::ReadJson(gameObjectPath);
	for (int x = 0; x < json["GameObjectComponentList"].size(); x++)
	{
		const int componentType = json["GameObjectComponentList"][x]["ComponentType"];
		switch (componentType)
		{
			case kTransform2DComponent:  LoadTransformComponent(json["GameObjectComponentList"][x], id, gameObjectPosition); break;
			case kInputComponent: LoadInputComponent(json["GameObjectComponentList"][x], id); break;
			case kSpriteComponent: LoadSpriteComponent(json["GameObjectComponentList"][x], id); break;
		}
	}
}

void GameObjectSystem::LoadTransformComponent(const nlohmann::json& json, GameObjectID id, const vec2& gameObjectPosition)
{
	Transform2DComponent transform2D;
	transform2D.GameObjectPosition = gameObjectPosition;
	transform2D.GameObjectRotation = vec2{ json["GameObjectRotation"][0], json["GameObjectRotation"][1] };
	transform2D.GameObjectScale = vec2{ json["GameObjectScale"][0], json["GameObjectScale"][1] };
	Transform2DComponentMap[id] = transform2D;
}

void GameObjectSystem::LoadInputComponent(const nlohmann::json& json, GameObjectID id)
{
	InputComponentMap[id] = InputComponent();
}

void GameObjectSystem::LoadSpriteComponent(const nlohmann::json& json, GameObjectID id)
{
	VkGuid vramId = VkGuid(json["VramId"].get<String>().c_str());
	levelSystem.SpriteList[id] = Sprite(id, vramId);
}

const GameObject& GameObjectSystem::FindGameObject(const GameObjectID& id)
{
	auto it = GameObjectMap.find(id);
	if (it != GameObjectMap.end())
	{
		return it->second;
	}
	throw std::out_of_range("GameObject not found for given GUID");
}

Transform2DComponent& GameObjectSystem::FindTransform2DComponent(const GameObjectID& id)
{
	auto it = Transform2DComponentMap.find(id);
	if (it != Transform2DComponentMap.end())
	{
		return it->second;
	}
	throw std::out_of_range("Transform2DComponent not found for given GUID");
}

const InputComponent& GameObjectSystem::FindInputComponent(const GameObjectID& id)
{
	auto it = InputComponentMap.find(id);
	if (it != InputComponentMap.end())
	{
		return it->second;
	}
	throw std::out_of_range("InputComponent not found for given GUID");
}

const Vector<GameObject> GameObjectSystem::GameObjectList()
{
	Vector<GameObject> GameObjectList;
	for (const auto& gameObjectMap : GameObjectMap)
	{
		GameObjectList.emplace_back(gameObjectMap.second);
	}
	return GameObjectList;
}

const Vector<Transform2DComponent> GameObjectSystem::Transform2DComponentList()
{
	Vector<Transform2DComponent> transform2DComponentList;
	for (const auto& transform2DComponent : Transform2DComponentMap)
	{
		transform2DComponentList.emplace_back(transform2DComponent.second);
	}
	return transform2DComponentList;
}

const Vector<InputComponent> GameObjectSystem::InputComponentList()
{
	Vector<InputComponent> inputComponentList;
	for (const auto& inputComponentMap : InputComponentMap)
	{
		inputComponentList.emplace_back(inputComponentMap.second);
	}
	return inputComponentList;
}

void GameObjectSystem::DestroyGameObject(const GameObjectID& gameObjectId)
{
//	GameObjectMap.erase(gameObjectId);
	Transform2DComponentMap.erase(gameObjectId);
	InputComponentMap.erase(gameObjectId);
}

void GameObjectSystem::DestroyGameObjects()
{
	for (auto& gameObject : GameObjectMap)
	{
		DestroyGameObject(gameObject.second.GameObjectId);
	}
}

File: VulkanGameEngineLevelEditor/Systems/GameObjectSystem.cs
using GlmSharp;
using Newtonsoft.Json;
using Silk.NET.SDL;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using VulkanGameEngineLevelEditor.GameEngineAPI;

namespace VulkanGameEngineLevelEditor.Systems
{
    public class GameObject
    {
        public int GameObjectId { get; set; }

        public GameObject()
        {

        }

        public GameObject(int gameObjectId)
        {
            GameObjectId = gameObjectId;
        }
    };

    public class GameObjectComponentLoader
    {
        public vec2 GameObjectPosition { get; set; }
        public vec2 GameObjectRotation { get; set; }
        public vec2 GameObjectScale { get; set; }
        public int ComponentType { get; set; }
        public Guid VramId { get; set; }
    }

    public class LoadGameObjectComponents
    {
        public List<GameObjectComponentLoader> GameObjectComponentList { get; set; }
    }

    public static class GameObjectSystem
    {
        public static Dictionary<int, GameObject> GameObjectMap { get; private set; } = new Dictionary<int, GameObject>();
        public static Dictionary<int, Transform2DComponent> Transform2DComponentMap { get; private set; } = new Dictionary<int, Transform2DComponent>();
        public static Dictionary<int, InputComponent> InputComponentMap { get; private set; } = new Dictionary<int, InputComponent>();

        public static void CreateGameObject(string name, List<ComponentTypeEnum> gameObjectComponentTypeList, Guid vramId, vec2 objectPosition)
        {
            int id = GameObjectMap.Count() + 1;
            GameObjectMap[id] = new GameObject(id);

            foreach (var component in gameObjectComponentTypeList)
            {
                switch (component)
                {
                    case  ComponentTypeEnum.kTransform2DComponent: Transform2DComponentMap[id] = new Transform2DComponent(objectPosition); break;
                    case ComponentTypeEnum.kInputComponent: InputComponentMap[id] = new InputComponent(id); break;
                    case ComponentTypeEnum.kSpriteComponent: SpriteSystem.AddSprite((uint)id, vramId); break;
                }
            }
        }

        public static void CreateGameObject(string gameObjectPath)
        {
            int id = GameObjectMap.Count() + 1;
            GameObjectMap[id] = new GameObject(id);

            string jsonContent = File.ReadAllText(gameObjectPath);
            LoadGameObjectComponents spriteVramJson = JsonConvert.DeserializeObject<LoadGameObjectComponents>(jsonContent);
            foreach(var component in spriteVramJson.GameObjectComponentList)
            {
                int componentType = component.ComponentType;
                switch ((ComponentTypeEnum)componentType)
                {
                    case ComponentTypeEnum.kTransform2DComponent: LoadTransformComponent(component, id, component.GameObjectPosition); break;
                    case ComponentTypeEnum.kInputComponent: LoadInputComponent(component, id); break;
                    case ComponentTypeEnum.kSpriteComponent: LoadSpriteComponent(component, (uint)id); break;
                }
            }
        }

        public static void LoadTransformComponent(GameObjectComponentLoader loader, int gameObjectId, vec2 gameObjectPosition)
        {
            Transform2DComponent transform2D = new Transform2DComponent();
            transform2D.GameObjectPosition = gameObjectPosition;
            transform2D.GameObjectRotation = loader.GameObjectRotation;
            transform2D.GameObjectScale = loader.GameObjectScale;
            Transform2DComponentMap[gameObjectId] = transform2D;
        }

        public static void LoadInputComponent(GameObjectComponentLoader loader, int gameObjectId)
        {
            InputComponentMap[gameObjectId] = new InputComponent();
        }

        public static void LoadSpriteComponent(GameObjectComponentLoader loader, uint gameObjectId)
        {
            SpriteSystem.AddSprite(gameObjectId, loader.VramId);
        }

        public static void DestroyGameObject(int gameObjectId)
        {
            //	GameObjectMap.erase(gameObjectId);
          //  Transform2DComponentMap.erase(gameObjectId);
           // InputComponentMap.erase(gameObjectId);
        }

        public static void DestroyGameObjects()
        {
            //for (auto & gameObject : GameObjectMap)
            //{
            //    DestroyGameObject(gameObject.second.GameObjectId);
            //}
        }
    }
}
